File: frontend/src/types/generacionEia.ts
/**
 * Tipos para Generación de EIA (Fase 4).
 */

// =============================================================================
// Enums
// =============================================================================

export type EstadoDocumentoEIA = 'borrador' | 'en_revision' | 'validado' | 'final'
export type FormatoExportacionEIA = 'pdf' | 'docx' | 'eseia_xml'
export type TipoValidacionEIA = 'contenido' | 'formato' | 'referencia' | 'completitud' | 'normativa'
export type SeveridadValidacion = 'error' | 'warning' | 'info'
export type EstadoObservacionEIA = 'pendiente' | 'corregida' | 'ignorada'
export type EstadoGeneracionEIA = 'pendiente' | 'generando' | 'completado' | 'error'

// =============================================================================
// Contenido de Capítulos
// =============================================================================

export interface ContenidoCapitulo {
  numero: number
  titulo: string
  contenido: string
  subsecciones: Record<string, unknown>
  figuras: string[]
  tablas: string[]
  referencias: string[]
}

export interface MetadatosDocumento {
  fecha_elaboracion?: string | null
  empresa_consultora?: string | null
  autores: string[]
  revisores: string[]
  version_doc?: string | null
  resumen_ejecutivo?: string | null
}

export interface EstadisticasDocumento {
  total_paginas: number
  total_palabras: number
  total_figuras: number
  total_tablas: number
  total_referencias: number
  capitulos_completados: number
  porcentaje_completitud: number
}

// =============================================================================
// Documento EIA
// =============================================================================

export interface DocumentoEIA {
  id: number
  proyecto_id: number
  titulo: string
  estado: EstadoDocumentoEIA
  version: number
  contenido_capitulos?: Record<string, ContenidoCapitulo> | null
  metadatos?: MetadatosDocumento | null
  validaciones?: Record<string, unknown> | null
  estadisticas?: EstadisticasDocumento | null
  created_at: string
  updated_at: string
}

export interface DocumentoEIAUpdate {
  titulo?: string
  estado?: EstadoDocumentoEIA
  contenido_capitulos?: Record<string, ContenidoCapitulo>
  metadatos?: MetadatosDocumento
}

// =============================================================================
// Versiones
// =============================================================================

export interface VersionEIA {
  id: number
  documento_id: number
  version: number
  cambios: string
  creado_por?: string | null
  created_at: string
}

export interface CrearVersionRequest {
  cambios: string
  creado_por?: string
}

// =============================================================================
// Generación
// =============================================================================

export interface CompilarDocumentoRequest {
  incluir_capitulos?: number[] | null
  regenerar_existentes: boolean
}

export interface GenerarCapituloRequest {
  capitulo_numero: number
  regenerar: boolean
  instrucciones_adicionales?: string | null
}

export interface RegenerarSeccionRequest {
  capitulo_numero: number
  seccion_codigo: string
  instrucciones: string
}

export interface ProgresoGeneracion {
  capitulo_numero: number
  titulo: string
  estado: EstadoGeneracionEIA
  progreso_porcentaje: number
  palabras_generadas: number
  tiempo_estimado_segundos?: number | null
  error?: string | null
}

export interface GeneracionResponse {
  documento_id: number
  capitulos_generados: number[]
  capitulos_con_error: number[]
  tiempo_total_segundos: number
  estadisticas: EstadisticasDocumento
}

// =============================================================================
// Validación
// =============================================================================

export interface ObservacionValidacion {
  id: number
  documento_id: number
  regla_id?: number | null
  capitulo_numero?: number | null
  seccion?: string | null
  tipo_observacion: string
  severidad: SeveridadValidacion
  mensaje: string
  sugerencia?: string | null
  contexto?: Record<string, unknown> | null
  estado: EstadoObservacionEIA
  created_at: string
  resuelta_at?: string | null
}

export interface ResultadoValidacion {
  documento_id: number
  total_observaciones: number
  observaciones_por_severidad: Record<string, number>
  observaciones_por_capitulo: Record<number, number>
  observaciones: ObservacionValidacion[]
  es_valido: boolean
  mensaje_resumen: string
}

export interface ValidacionRequest {
  incluir_capitulos?: number[] | null
  nivel_severidad_minima: SeveridadValidacion
}

// =============================================================================
// Exportación
// =============================================================================

export interface ConfiguracionPDF {
  incluir_portada: boolean
  incluir_indice: boolean
  incluir_figuras: boolean
  incluir_tablas: boolean
  tamano_pagina: string
  orientacion: 'portrait' | 'landscape'
  margen_cm: number
  fuente: string
  tamano_fuente: number
  logo_empresa?: string | null
}

export interface ConfiguracionDOCX {
  incluir_portada: boolean
  incluir_indice: boolean
  incluir_figuras: boolean
  incluir_tablas: boolean
  estilo_documento: string
  logo_empresa?: string | null
}

export interface ConfiguracionESEIA {
  incluir_anexos_digitales: boolean
  version_esquema: string
}

export interface ExportacionRequest {
  formato: FormatoExportacionEIA
  configuracion?: Record<string, unknown> | null
}

export interface ExportacionEIA {
  id: number
  documento_id: number
  formato: FormatoExportacionEIA
  archivo_path?: string | null
  tamano_bytes?: number | null
  generado_exitoso: boolean
  error_mensaje?: string | null
  created_at: string
  url_descarga?: string | null
}

// =============================================================================
// Documento Completo (con relaciones)
// =============================================================================

export interface DocumentoCompletoResponse {
  documento: DocumentoEIA
  versiones: VersionEIA[]
  exportaciones: ExportacionEIA[]
  observaciones_pendientes: ObservacionValidacion[]
  estadisticas: EstadisticasDocumento
  progreso_capitulos: ProgresoGeneracion[]
}

// =============================================================================
// Labels y constantes
// =============================================================================

export const ESTADO_DOCUMENTO_LABELS: Record<EstadoDocumentoEIA, string> = {
  borrador: 'Borrador',
  en_revision: 'En Revisión',
  validado: 'Validado',
  final: 'Final',
}

export const ESTADO_DOCUMENTO_COLORS: Record<EstadoDocumentoEIA, string> = {
  borrador: 'badge-ghost',
  en_revision: 'badge-warning',
  validado: 'badge-info',
  final: 'badge-success',
}

export const FORMATO_EXPORTACION_LABELS: Record<FormatoExportacionEIA, string> = {
  pdf: 'PDF',
  docx: 'Word (DOCX)',
  eseia_xml: 'e-SEIA (XML)',
}

export const FORMATO_EXPORTACION_ICONS: Record<FormatoExportacionEIA, string> = {
  pdf: 'heroicons:document-text',
  docx: 'heroicons:document',
  eseia_xml: 'heroicons:code-bracket',
}

export const SEVERIDAD_LABELS: Record<SeveridadValidacion, string> = {
  error: 'Error',
  warning: 'Advertencia',
  info: 'Información',
}

export const SEVERIDAD_COLORS: Record<SeveridadValidacion, string> = {
  error: 'badge-error',
  warning: 'badge-warning',
  info: 'badge-info',
}

export const SEVERIDAD_ICONS: Record<SeveridadValidacion, string> = {
  error: 'heroicons:x-circle',
  warning: 'heroicons:exclamation-triangle',
  info: 'heroicons:information-circle',
}

export const ESTADO_OBSERVACION_LABELS: Record<EstadoObservacionEIA, string> = {
  pendiente: 'Pendiente',
  corregida: 'Corregida',
  ignorada: 'Ignorada',
}

export const ESTADO_GENERACION_LABELS: Record<EstadoGeneracionEIA, string> = {
  pendiente: 'Pendiente',
  generando: 'Generando...',
  completado: 'Completado',
  error: 'Error',
}

export const ESTADO_GENERACION_COLORS: Record<EstadoGeneracionEIA, string> = {
  pendiente: 'badge-ghost',
  generando: 'badge-warning',
  completado: 'badge-success',
  error: 'badge-error',
}

// Títulos de los 11 capítulos del EIA
export const CAPITULOS_EIA: Record<number, string> = {
  1: 'Índice',
  2: 'Resumen Ejecutivo',
  3: 'Descripción del Proyecto',
  4: 'Plan de Cumplimiento de Legislación Ambiental',
  5: 'Permisos Ambientales Sectoriales',
  6: 'Línea de Base',
  7: 'Evaluación de Impacto Ambiental',
  8: 'Plan de Medidas de Mitigación, Reparación y Compensación',
  9: 'Plan de Prevención de Contingencias y Emergencias',
  10: 'Plan de Seguimiento Ambiental',
  11: 'Ficha Resumen del EIA',
}

File: backend/app/schemas/generacion_eia.py
"""
Schemas Pydantic para Generación de EIA (Fase 4).
Define tipos para compilación, generación, validación y exportación de documentos EIA.
"""
from datetime import datetime
from typing import Optional, List, Dict, Any
from enum import Enum
from pydantic import BaseModel, Field


# =============================================================================
# Enums
# =============================================================================

class EstadoDocumentoEnum(str, Enum):
    """Estados posibles de un documento EIA."""
    BORRADOR = "borrador"
    EN_REVISION = "en_revision"
    VALIDADO = "validado"
    FINAL = "final"


class FormatoExportacionEnum(str, Enum):
    """Formatos de exportación disponibles."""
    PDF = "pdf"
    DOCX = "docx"
    ESEIA_XML = "eseia_xml"


class TipoValidacionEnum(str, Enum):
    """Tipos de validación SEA."""
    CONTENIDO = "contenido"
    FORMATO = "formato"
    REFERENCIA = "referencia"
    COMPLETITUD = "completitud"
    NORMATIVA = "normativa"


class SeveridadEnum(str, Enum):
    """Niveles de severidad de observaciones."""
    ERROR = "error"
    WARNING = "warning"
    INFO = "info"


class EstadoObservacionEnum(str, Enum):
    """Estados de observaciones de validación."""
    PENDIENTE = "pendiente"
    CORREGIDA = "corregida"
    IGNORADA = "ignorada"


# =============================================================================
# Documento EIA
# =============================================================================

class ContenidoCapitulo(BaseModel):
    """Contenido de un capítulo individual."""
    numero: int = Field(..., ge=1, le=11)
    titulo: str
    contenido: str
    subsecciones: Dict[str, Any] = Field(default_factory=dict)
    figuras: List[str] = Field(default_factory=list)
    tablas: List[str] = Field(default_factory=list)
    referencias: List[str] = Field(default_factory=list)


class MetadatosDocumento(BaseModel):
    """Metadatos del documento EIA."""
    fecha_elaboracion: Optional[datetime] = None
    empresa_consultora: Optional[str] = None
    autores: List[str] = Field(default_factory=list)
    revisores: List[str] = Field(default_factory=list)
    version_doc: Optional[str] = None
    resumen_ejecutivo: Optional[str] = None


class EstadisticasDocumento(BaseModel):
    """Estadísticas del documento."""
    total_paginas: int = 0
    total_palabras: int = 0
    total_figuras: int = 0
    total_tablas: int = 0
    total_referencias: int = 0
    capitulos_completados: int = 0
    porcentaje_completitud: float = 0.0


class DocumentoEIABase(BaseModel):
    """Base para documento EIA."""
    proyecto_id: int
    titulo: str = Field(..., min_length=10, max_length=500)
    estado: EstadoDocumentoEnum = EstadoDocumentoEnum.BORRADOR


class DocumentoEIACreate(DocumentoEIABase):
    """Schema para crear documento EIA."""
    metadatos: Optional[MetadatosDocumento] = None


class DocumentoEIAUpdate(BaseModel):
    """Schema para actualizar documento EIA."""
    titulo: Optional[str] = Field(None, min_length=10, max_length=500)
    estado: Optional[EstadoDocumentoEnum] = None
    contenido_capitulos: Optional[Dict[str, ContenidoCapitulo]] = None
    metadatos: Optional[MetadatosDocumento] = None


class DocumentoEIAResponse(DocumentoEIABase):
    """Respuesta de documento EIA."""
    id: int
    version: int
    contenido_capitulos: Optional[Dict[str, Any]] = None
    metadatos: Optional[Dict[str, Any]] = None
    validaciones: Optional[Dict[str, Any]] = None
    estadisticas: Optional[Dict[str, Any]] = None
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


# =============================================================================
# Versiones
# =============================================================================

class VersionEIACreate(BaseModel):
    """Schema para crear versión."""
    cambios: str = Field(..., min_length=10)
    creado_por: Optional[str] = None


class VersionEIAResponse(BaseModel):
    """Respuesta de versión EIA."""
    id: int
    documento_id: int
    version: int
    cambios: str
    creado_por: Optional[str]
    created_at: datetime

    class Config:
        from_attributes = True


# =============================================================================
# Generación
# =============================================================================

class GenerarCapituloRequest(BaseModel):
    """Request para generar un capítulo específico."""
    capitulo_numero: int = Field(..., ge=1, le=11)
    regenerar: bool = False
    instrucciones_adicionales: Optional[str] = None


class RegenerarSeccionRequest(BaseModel):
    """Request para regenerar una sección específica."""
    capitulo_numero: int = Field(..., ge=1, le=11)
    seccion_codigo: str
    instrucciones: str = Field(..., min_length=10)


class CompilarDocumentoRequest(BaseModel):
    """Request para compilar documento completo."""
    incluir_capitulos: Optional[List[int]] = None  # Si es None, incluye todos
    regenerar_existentes: bool = False


class ProgresoGeneracion(BaseModel):
    """Progreso de generación por capítulo."""
    capitulo_numero: int
    titulo: str
    estado: str  # pendiente, generando, completado, error
    progreso_porcentaje: int = 0
    palabras_generadas: int = 0
    tiempo_estimado_segundos: Optional[int] = None
    error: Optional[str] = None


class GeneracionResponse(BaseModel):
    """Respuesta de proceso de generación."""
    documento_id: int
    capitulos_generados: List[int]
    capitulos_con_error: List[int]
    tiempo_total_segundos: float
    estadisticas: EstadisticasDocumento


# =============================================================================
# Validación
# =============================================================================

class ReglaValidacionSEABase(BaseModel):
    """Base para regla de validación."""
    codigo_regla: str = Field(..., min_length=3, max_length=50)
    descripcion: str
    tipo_validacion: TipoValidacionEnum
    criterio: Dict[str, Any]
    mensaje_error: str
    mensaje_sugerencia: Optional[str] = None
    severidad: SeveridadEnum = SeveridadEnum.WARNING


class ReglaValidacionSEACreate(ReglaValidacionSEABase):
    """Schema para crear regla de validación."""
    tipo_proyecto_id: Optional[int] = None
    capitulo_numero: Optional[int] = None


class ReglaValidacionSEAResponse(ReglaValidacionSEABase):
    """Respuesta de regla de validación."""
    id: int
    tipo_proyecto_id: Optional[int]
    capitulo_numero: Optional[int]
    activo: bool
    created_at: datetime

    class Config:
        from_attributes = True


class ObservacionValidacionBase(BaseModel):
    """Base para observación de validación."""
    tipo_observacion: str
    severidad: SeveridadEnum
    mensaje: str
    sugerencia: Optional[str] = None


class ObservacionValidacionCreate(ObservacionValidacionBase):
    """Schema para crear observación."""
    documento_id: int
    regla_id: Optional[int] = None
    capitulo_numero: Optional[int] = None
    seccion: Optional[str] = None
    contexto: Optional[Dict[str, Any]] = None


class ObservacionValidacionResponse(ObservacionValidacionBase):
    """Respuesta de observación."""
    id: int
    documento_id: int
    regla_id: Optional[int]
    capitulo_numero: Optional[int]
    seccion: Optional[str]
    contexto: Optional[Dict[str, Any]]
    estado: EstadoObservacionEnum
    created_at: datetime
    resuelta_at: Optional[datetime]

    class Config:
        from_attributes = True


class ValidacionRequest(BaseModel):
    """Request para validar documento."""
    incluir_capitulos: Optional[List[int]] = None
    nivel_severidad_minima: SeveridadEnum = SeveridadEnum.INFO


class ResultadoValidacion(BaseModel):
    """Resultado de validación de documento."""
    documento_id: int
    total_observaciones: int
    observaciones_por_severidad: Dict[str, int]
    observaciones_por_capitulo: Dict[int, int]
    observaciones: List[ObservacionValidacionResponse]
    es_valido: bool
    mensaje_resumen: str


# =============================================================================
# Exportación
# =============================================================================

class ConfiguracionPDF(BaseModel):
    """Configuración para exportación PDF."""
    incluir_portada: bool = True
    incluir_indice: bool = True
    incluir_figuras: bool = True
    incluir_tablas: bool = True
    tamano_pagina: str = "A4"
    orientacion: str = "portrait"  # portrait, landscape
    margen_cm: float = 2.5
    fuente: str = "Arial"
    tamano_fuente: int = 11
    logo_empresa: Optional[str] = None


class ConfiguracionDOCX(BaseModel):
    """Configuración para exportación DOCX."""
    incluir_portada: bool = True
    incluir_indice: bool = True
    incluir_figuras: bool = True
    incluir_tablas: bool = True
    estilo_documento: str = "SEA_Estandar"
    logo_empresa: Optional[str] = None


class ConfiguracionESEIA(BaseModel):
    """Configuración para exportación e-SEIA XML."""
    incluir_anexos_digitales: bool = True
    version_esquema: str = "2.0"


class ExportacionRequest(BaseModel):
    """Request para exportar documento."""
    formato: FormatoExportacionEnum
    configuracion: Optional[Dict[str, Any]] = None


class ExportacionResponse(BaseModel):
    """Respuesta de exportación."""
    id: int
    documento_id: int
    formato: FormatoExportacionEnum
    archivo_path: Optional[str]
    tamano_bytes: Optional[int]
    generado_exitoso: bool
    error_mensaje: Optional[str]
    created_at: datetime
    url_descarga: Optional[str] = None

    class Config:
        from_attributes = True


# =============================================================================
# Templates
# =============================================================================

class TemplateCapituloBase(BaseModel):
    """Base para template de capítulo."""
    capitulo_numero: int = Field(..., ge=1, le=11)
    titulo: str = Field(..., min_length=5, max_length=200)
    template_prompt: str = Field(..., min_length=50)
    instrucciones_generacion: Optional[str] = None


class TemplateCapituloCreate(TemplateCapituloBase):
    """Schema para crear template."""
    tipo_proyecto_id: int
    estructura_esperada: Optional[Dict[str, Any]] = None
    ejemplos: Optional[str] = None


class TemplateCapituloResponse(TemplateCapituloBase):
    """Respuesta de template."""
    id: int
    tipo_proyecto_id: int
    estructura_esperada: Optional[Dict[str, Any]]
    ejemplos: Optional[str]
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


# =============================================================================
# Respuestas Compuestas
# =============================================================================

class DocumentoCompletoResponse(BaseModel):
    """Respuesta completa de documento con todas sus relaciones."""
    documento: DocumentoEIAResponse
    versiones: List[VersionEIAResponse]
    exportaciones: List[ExportacionResponse]
    observaciones_pendientes: List[ObservacionValidacionResponse]
    estadisticas: EstadisticasDocumento
    progreso_capitulos: List[ProgresoGeneracion]
